mlir/lib/Dialect/Tosa/Transforms/SinkInputOpsThroughConcat.cpp

```diff
@@ -0,0 +1,430 @@
+
+protected:
+  // check that the concat happens on another axis than the one used by matmul
+  LogicalResult doExtraChecks(tosa::MatMulOp op, tosa::ConcatOp concat,
+                              PatternRewriter &rewriter) const override {
+    if (concat.getAxis() != 0)
+      return rewriter.notifyMatchFailure(
+          concat, "Matmul concat about different axis not yet supported");
+    return success();
+  }
+};
+
+struct SinkReshapeOp : public SinkSpecificOp<tosa::ReshapeOp> {
+  using SinkSpecificOp<tosa::ReshapeOp>::SinkSpecificOp;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    auto reshapeOrError = doGenericChecks(concatOp, rewriter);
+    if (failed(reshapeOrError))
+      return reshapeOrError;
+    auto reshape = *reshapeOrError;
+
+    assert(reshape->getNumOperands() == 1 &&
+           "shape must have exactly one operand");
+    const auto opType = reshape->getOperandTypes().front();
+    const auto tenType = dyn_cast<ShapedType>(opType);
+    if (!tenType || !tenType.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          concatOp, "Dynamic shapes for reshapes are not supported.");
+    const auto oldShape = tenType.getShape();
+    const auto newShape = reshape.getNewShape();
+
+    const auto newConcatAxis = concatOp.getAxis();
+    int64_t newPrefixProduct = 1;
+    for (size_t i = 0; i < newConcatAxis; ++i) {
+      newPrefixProduct *= newShape[i];
+    }
+
+    int64_t oldPrefixProduct = 1;
+    size_t oldConcatAxis = oldShape.size(); // default to invalid value
+    size_t i = 0;
+    assert(oldShape.size() > 0 &&
+           "Input sensor of rank 0 does not make sense here.");
+    do {
+      if (oldPrefixProduct == newPrefixProduct) {
+        oldConcatAxis = i;
+        break;
+      }
+      oldPrefixProduct *= oldShape[i];
+      i++;
+    } while (i < oldShape.size() - 1);
+
+    if (oldConcatAxis == oldShape.size())
+      return rewriter.notifyMatchFailure(
+          concatOp, "Sinking reshape not possible. No compatible dimension for "
+                    "concat axis found.");
+
+    SmallVector<Value> concatOperands;
+    for (auto val : concatOp.getOperands()) {
+      auto *producerOp = val.getDefiningOp();
+      assert(producerOp != nullptr &&
+             "Previous check about null already happened.");
+      for (auto val : producerOp->getOperands()) {
+        concatOperands.emplace_back(val);
+      }
+    }
+    auto concatNew = rewriter.create<tosa::ConcatOp>(
+        concatOp.getLoc(), concatOperands, oldConcatAxis);
+    // calculate new shape for reshape by combining the shape of the concat with
+    // the remaining prefix of the old reshape
+    auto concatResult = concatNew->getResultTypes();
+    auto concatType = concatResult.front();
+    auto concatShapeT = dyn_cast<ShapedType>(concatType);
+    assert(concatShapeT && "op and thus concat must have a shape");
```

Comment:
You can accomplish the same with:

```suggestion
    auto concatShapeT = cast<ShapedType>(concatType);
```

Cast asserts internally